output/mistral_fluentbit/mistral_fluentbit_tcp.c

```diff
@@ -0,0 +1,246 @@
+        mistral_err("Server name too long!\n");
+
+        /* This is thread safe because the connectivity thread hasn#t been created yet*/
+        ctx->state = failed;
+        return (-1);
+    }
+
+    mistral_fluentbit_thread_cb_s *ptr_data = (mistral_fluentbit_thread_cb_s *)malloc(
+        sizeof(mistral_fluentbit_thread_cb_s));
+    if (ptr_data == NULL) {
+        /* This is thread safe because the connectivity thread hasn't been created yet*/
+        ctx->state = failed;
+        return (-1);
+    }
+
+    strcpy(ptr_data->server_name, servername);
+    ptr_data->port = port;
+
+    ptr_data->cb = func;
+    ptr_data->data = args;
+    ptr_data->ctx = ctx;
+
+    if ((ret = pthread_create(&thread_id, NULL, 
+        mistral_tcp_connection_thread, (void *)ptr_data)) != 0)
+    {
+        mistral_err("pthread_create() error %d\n", ret);
+        ctx->state = failed;
+        return (-1);
+    }
+
+    if (pthread_detach(thread_id) != 0) {
+        mistral_err("pthread_detach() error\n");
+        return (-1);
+    }
+
+    pthread_mutex_lock(&ctx->mutex);
+    ctx->thread_id = thread_id;
+    pthread_mutex_unlock(&ctx->mutex);
+
+    return ret;
+}
+
+int mistral_fluentbit_send(mistral_fluentbit_tcp_ctx_s *ctx, const char *buff, int len)
+{
+    int fd = -1;
+    mistral_fluentbit_tcp_state_t state;
+
+    pthread_mutex_lock(&ctx->mutex);
+    fd = ctx->fd;
+    state = ctx->state;
+    pthread_mutex_unlock(&ctx->mutex);
+
+    if (state != connected) {
+        return (-1);
+    }
+
+    int ret =  write(fd, buff, len);
+
+    if (ret < 0) {
+        mistral_err("mistral_fluentbit_send <%d>!\n", ret);
+    }
+
+    return ret;
+}
+
+static int mistral_tcp_connect(int sockfd, struct sockaddr_in *serveraddr, int retries, useconds_t
+                               usec)
+{
+    int iterations = 0;
+    int ret = -1;
+
+    while (iterations < retries) {
+        ret = connect(sockfd, (const struct sockaddr *)serveraddr, sizeof(struct sockaddr_in));
+        /* Connect: create a connection to the Fluet Bit server */
+        if (ret < 0) {
+            mistral_err("Error connecting socket %d\n", sockfd);
+            iterations++;
+            usleep(iterations * usec);
+        } else {
+            goto done;
+        }
+    }
+
+done:
+    return ret;
+}
+
+static void *mistral_tcp_connection_thread(void *arg)
+{
+    int sockfd = -1;
+    struct sockaddr_in serveraddr;
+    struct hostent *server = NULL;
+    char buff[MISTRAL_TCP_BUFFER_SIZE];
+
+    mistral_fluentbit_thread_cb_s *ptr_data = (mistral_fluentbit_thread_cb_s *)arg;
+    mistral_fluentbit_tcp_ctx_s *tcp_ctx = ptr_data->ctx;
+
+    /* Socket: create the socket */
+    sockfd = socket(AF_INET, SOCK_STREAM, 0);
+    if (sockfd < 0) {
+        mistral_err("Error opening socket %d\n", sockfd);
+        goto failed;
+    }
+
+    /* Gethostbyname: get the server's DNS entry */
+    server = gethostbyname(ptr_data->server_name);
```

Comment:
Stupid question - does this work with both a hostname and an IP address?

Reply:
It works with both hostname and IP address.